fix: #74 drafty should correctly parse unicode text with use of graphemes

dilshans2k opened the pull request:
- drafty should correctly parse unicode text with use of graphemes
- add more drafty ut

dilshans2k committed:
fix: #74 drafty should correctly parse unicode text with use of graphemes

- add more drafty ut

or-else commented:
Could you please sign the CLA: https://github.com/tinode/chat/blob/master/docs/CLA.md

Thanks!

or-else commented on src/drafty.js:
```diff
@@ -2452,12 +2495,15 @@ function shortenTree(tree, limit, tail) {
+        node.text = graphemes
+          .slice(0, limit)
+          .map((segment) => segment.segment)
+          .join("") + tail;
```
Code style: please use `'` (single quotes) for strings which require no localization.

or-else commented on src/drafty.js:
```diff
@@ -2262,7 +2300,9 @@ function spansToTree(parent, text, start, end, spans) {
-        text: text.substring(start, span.start)
+        text: graphemes.slice(start, span.start)
+          .map(segment => segment.segment)
+          .join("")
```
Code style: please use ' (single quotes) for strings which require no localization.

or-else commented on src/drafty.js:
```diff
@@ -2236,11 +2272,13 @@ function addNode(parent, n) {
-        text: text.substring(start, end)
+        text: graphemes.slice(start, end)
+          .map(segment => segment.segment)
+          .join("")
```
Code style: please use ' (single quotes) for strings which require no localization.

or-else commented on src/drafty.js:
```diff
@@ -2295,14 +2335,17 @@ function spansToTree(parent, text, start, end, spans) {
+      text: graphemes
+        .slice(start, end)
+        .map((segment) => segment.segment)
+        .join("")
```
Code style: please use ' (single quotes) for strings which require no localization.

or-else commented on src/drafty.js:
```diff
@@ -2633,6 +2679,63 @@ function copyEntData(data, light, allow) {
 
+// Returns true if object is empty, if undefined returns true
+function isEmptyObject(obj) {
+  return Object.keys(obj ?? {}).length === 0;
```
We do not use `===` unless required. Please use `==` here.

or-else commented on src/drafty.js:
```diff
@@ -2633,6 +2679,63 @@ function copyEntData(data, light, allow) {
+  let charIndex = 0;
+
+  // Iterate over the grapheme clusters
+  for (let {
```
`for (const {`

or-else commented on src/drafty.js:
```diff
@@ -597,28 +599,61 @@ Drafty.parse = function(content) {
+    if (result.fmt.length) {
+      const segments = segmenter.segment(result.txt);
+      for (const ele of result.fmt) {
+        ({
```
You use `const {` everywhere else. I suggest to choose one and stick to it: either `const {` or `({`. I like `const {` more because it's more readable.

dilshans2k replied:
We cannot use `const {` notation here because no new variable is declared. We are assigning the value directly to field of `ele` object. 
If it is required that we use `const { ` notation, something like this would work

```
const {
    at: correctAt,
    len: correctLen
  } = getCorrectLengthsWhenTreatedAsGrapheme(ele, segments, result.txt);
ele.at = correctAt;
ele.len = correctLen;
```

or-else commented on src/drafty.js:
```diff
@@ -597,28 +599,61 @@ Drafty.parse = function(content) {
+        segments = segmenter.segment(block.txt);
+        result.fmt = result.fmt.concat(
+          block.fmt.map((s) => {
+            const {
```
You use `({` above. I suggest to choose one and stick to it: either `const {` or `({`. I like `const {` more.

or-else commented on src/drafty.js:
```diff
@@ -2633,6 +2679,63 @@ function copyEntData(data, light, allow) {
+// denotes the position of char in string in a grapheme array(created from that string)
+// Eg: string: "Hi👋🏼Hi" -> [0,1,2,2,2,2,3,4]
+function getGraphemeIndices(graphemes) {
+  let result = [];
```
`const result = [];`

or-else commented on src/drafty.js:
```diff
@@ -2187,7 +2222,8 @@ function draftyToTree(doc) {
   });
 
-  let tree = spansToTree({}, txt, 0, txt.length, spans);
+  const graphemes = getGraphemesFromString(txt)
```
Terminating semicolon is missing.

dilshans2k commented on src/drafty.js:
```diff
@@ -67,6 +67,8 @@ const ALLOWED_ENT_FIELDS = ['act', 'height', 'duration', 'incoming', 'mime', 'na
   'ref', 'size', 'state', 'url', 'val', 'width'
 ];
 
+const segmenter = new Intl.Segmenter();
```
> This will fail in Firefox. Do you want to add dependency on https://www.npmjs.com/package/@formatjs/intl-segmenter?

Yes, you are right. But I thought `formatjs/intl-segmenter` dependency would go in webapp and polyfill will be enabled there too, instead of sdk.

> FormatJS can be useful elsewhere. There are some error messages in the package which are not localized. I would use FormatJS to localize them.

Can you please elaborate? I am unable to understand which errors you are referring to.

or-else replied:
1. Webapp imports `'@formatjs/intl'`, not `@formatjs/intl-segmenter/polyfill`. I think it's going to fail in FF without doing something like this: https://formatjs.io/docs/polyfills/intl-segmenter/
2. The SDK is currently not internationalized. It has quite a few hardcoded English strings. It would be nice to internationalize it too. I would use `FormatJS` for that.

dilshans2k replied:
I have imported formatjs's `intl-segmenter` `polyfill` in `tinode-js` instead of `webapp`.

Certainly, sdk can be internationalized in another PR, but it is out of scope of this one.

or-else commented:
Any update on this? When do you expect to finish this PR?

or-else commented:
I this PR abandoned?

dilshans2k commented:
I was just busy with something else. I will complete it soon.

dilshans2k committed:
refactor: #74 some refactors

- use '' single quotes for non-localizable strings
- use const instead of let where required
- use == instead of === where required

dilshans2k committed:
feat: #74 import intl-segmenter polfill from formatjs

- polyfill is required because firefox hasnt implemented intl segmenter yet

or-else commented:
Thanks!

or-else commented:
The inclusion of @formatjs/segmenter increases the size of the prod bundle more that 4 times, from 80KB to 380KB. Devel goes from 189KB to 730KB. I won't be able to release it like this.
